java/ql/lib/semmle/code/java/frameworks/android/ExternalStorage.qll

```diff
@@ -0,0 +1,50 @@
+/** Provides definitions for working with uses of Android external storage */
+
+import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.ExternalFlow
+
+private class ExternalStorageDirSourceModel extends SourceModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        //"package;type;overrides;name;signature;ext;spec;kind"
+        "android.content;Context;true;getExternalFilesDir;(String);;ReturnValue;android-external-storage-dir",
+        "android.content;Context;true;getExternalFilesDirs;(String);;ReturnValue;android-external-storage-dir",
+        "android.content;Context;true;getExternalCacheDir;();;ReturnValue;android-external-storage-dir",
+        "android.content;Context;true;getExternalCacheDirs;();;ReturnValue;android-external-storage-dir",
+        "android.os;Environment;false;getExternalStorageDirectory;();;ReturnValue;android-external-storage-dir",
+        "android.os;Environment;false;getExternalStoragePublicDirectory;(String);;ReturnValue;android-external-storage-dir",
+      ]
+  }
+}
+
+private predicate externalStorageFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+  DataFlow::localFlowStep(node1, node2)
+  or
+  exists(ConstructorCall c | c.getConstructedType() instanceof TypeFile |
+    node1.asExpr() = c.getArgument(0) and
+    node2.asExpr() = c
+  )
+  or
+  node2.asExpr().(ArrayAccess).getArray() = node1.asExpr()
+  or
+  node2.asExpr().(FieldRead).getField().getInitializer() = node1.asExpr()
+}
+
+private predicate externalStorageFlow(DataFlow::Node node1, DataFlow::Node node2) {
+  externalStorageFlowStep*(node1, node2)
+}
+
+/**
+ * Holds if `n` is a node that reads the contents of an external file in Android.
+ * This is controlable by third-party applications, so is treated as a remote flow source.
+ */
+predicate androidExternalStorageSource(DataFlow::Node n) {
+  exists(ConstructorCall fInp, DataFlow::Node externalDir |
+    fInp.getConstructedType().hasQualifiedName("java.io", "FileInputStream") and
+    n.asExpr() = fInp and
+    sourceNode(externalDir, "android-external-storage-dir") and
+    externalStorageFlow(externalDir, DataFlow::exprNode(fInp.getArgument(0)))
+  )
```

Comment:
I'd suggest using `FileReadExpr` from `FileReadWrite.qll` instead of only the `FileInputStream` constructor to catch more ways of reading files.

```suggestion
predicate androidExternalStorageSource(DataFlow::Node n) {
  exists(DataFlow::Node externalDir |
    sourceNode(externalDir, "android-external-storage-dir") and
    externalStorageFlow(externalDir, DataFlow::exprNode(n.asExpr().(FileReadExpr).getFileVarAccess()))
  )
```

Although I think we'd need to change the `FileReadExpr::getFileVarAccess` predicate to be something like `getFileExpr`, to not restrict ourselves to `VarAccess`es. That impacts the `ReadingFromWorldWritableFile.ql` query, so we can handle that in a follow-up PR if you prefer.

Reply:
It looks like `FileReadExpr` includes nested calls, which don't exactly correspond to dataflow and thus could cause FPs. 
For example, in the following: 
```java
String foo(File f) {
  this.is = FileInputStream(f);
  return "something else";
}

void bar(Contex ctx) {
  String s = foo(new File(ctx.getExternalFilesDir(null), "file.txt"));
  sink(s);
}
```
then the call to `foo` is considered a `FileReadExpr` but should not be tainted.

Instead perhaps we could define a variant of `FileReadExpr` that doesn't look through nested calls.

Reply:
Good catch. Yes, my main concern was not repeating a list of methods that read from files if possible.

Maybe the current `FileReadExpr` (which I think needs a name change to indicate its recursive nature) could build upon the new version, which wouldn't look through nested calls and could be reused by other libraries interested in file-reading methods.